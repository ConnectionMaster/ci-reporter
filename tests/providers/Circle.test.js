const Circle = require('../../lib/providers/Circle')
const nock = require('nock')
const fs = require('fs')
const path = require('path')

const readFile = file => fs.readFileSync(path.join(__dirname, '..', 'fixtures', 'circle', file), 'utf8')

describe('Circle', () => {
  describe('static get ctx()', () => {
    it('returns the correct status context string', () => {
      expect(Circle.ctx).toBe('ci/circleci')
    })
  })

  describe('buildUri', () => {
    it('creates the correct URI', () => {
      const circle = new Circle()
      const repo = { owner: 'JasonEtco', repo: 'ci-reporter' }
      expect(circle.buildUri(repo, 317090494)).toBe('https://circleci.com/api/v1.1/project/github/JasonEtco/ci-reporter/317090494')
    })
  })

  describe('parseLog', () => {
    const log = readFile('log.txt')
    it('returns the correct log string', () => {
      const circle = new Circle()
      const actual = circle.parseLog(log)
      expect(actual).toMatchSnapshot()
    })
  })

  describe('serialize', () => {
    let circle

    beforeEach(() => {
      const build = readFile('build.json')
      const output = readFile('output.json')

      nock('https://circleci.com')
        .get('/api/v1.1/project/github/JasonEtco/todo/5').reply(200, build)
        .get('/fake-output-url').reply(200, output)

      circle = new Circle({
        payload: {
          target_url: 'https://circleci.com/gh/JasonEtco/todo/5?utm_source=github_status&utm_medium=notification'
        },
        repo: () => ({ owner: 'JasonEtco', repo: 'todo' })
      })
    })

    it('returns the correct body string', async () => {
      const res = await circle.serialize()
      expect(res.number).toBe(1)
      expect(res.body).toMatchSnapshot()
    })

    it('returns false if the status is not on a PR', async () => {
      nock.cleanAll()
      nock('https://circleci.com')
        .get('/api/v1.1/project/github/JasonEtco/todo/5').reply(200, readFile('commit.json'))

      const res = await circle.serialize()
      expect(res).toBeFalsy()
    })

    // it('returns the correct body string with multiple jobs', async () => {
    //   nock.cleanAll()
    //   nock('https://api.circle-ci.org')
    //     .get('/builds/123').reply(200, JSON.stringify({
    //       build: { pull_request_number: 1 },
    //       jobs: [{ id: 1234, number: 1, state: 'failed' }, { id: 12345, number: 1, state: 'failed' }]
    //     }))
    //     .get('/jobs/1234/log').reply(200, logFile)
    //     .get('/jobs/12345/log').reply(200, logFile)

    //   const res = await circle.serialize()
    //   expect(res.number).toBe(1)
    //   expect(res.body).toMatchSnapshot()
    // })
  })
})
