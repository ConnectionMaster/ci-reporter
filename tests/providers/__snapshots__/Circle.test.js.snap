// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle parseLog returns the correct log string 1`] = `
"> failing-repo@1.0.0 test /home/circleci/repo
> echo \\"Error: no test specified\\" && exit 1

Error: no test specified"
`;

exports[`Circle serialize returns the correct body string 1`] = `
":sparkles: Good work on this PR so far! :sparkles: Unfortunately, the [Circle CI build](https://circleci.com/gh/JasonEtco/todo/5?utm_source=github_status&utm_medium=notification) is failing. Here's the output:
\`\`\`

\`\`\`
I'm sure you can fix it! If you need help, don't hesitate to ask a maintainer of the project!"
`;
